myldr/boot.c

```diff
@@ -204,6 +264,66 @@ typedef BOOL (WINAPI *pALLOW)(DWORD);
     /* extract embedded_files[0] (i.e. the custom Perl interpreter) 
      * into stmpdir (but under the same basename as argv[0]) */
     my_prog = par_findprog(argv[0], par_getenv("PATH"));
+
+#ifdef __MACH__
+    /* Detect if FAT binary */
+    FILE *obj_file = fopen(my_prog, "rb");
+    uint32_t magic = read_magic(obj_file, 0);
+    fclose(obj_file);
+    if (magic == FAT_CIGAM || magic == FAT_MAGIC) {
+
+      /* Create seperate dir for extracted thin binary*/
+      char *ftmpdir = malloc(strlen(stmpdir)+4);
+      ftmpdir = par_mktmpdir( argv );
+      strcat(ftmpdir, "_FAT");
+      sanitise_tmp(ftmpdir);
+      rc = my_mkdir(ftmpdir, 0700);
+      if ( rc == -1 && errno != EEXIST) {
+	die("%s: creation of MacOS extracted thin binary private cache subdirectory %s failed (errno= %i)\n", 
+            argv[0], ftmpdir, errno);
+      }
+      
+      /* Get architecture name */
+      size_t size;
+      sysctlbyname("hw.machine", NULL, &size, NULL, 0);
+      char *arch = malloc(size);
+      sysctlbyname("hw.machine", arch, &size, NULL, 0);
+
+      /* Detect if lipo exists, if not die */
+      struct stat buffer;
+      int exist = stat("/usr/bin/lipo", &buffer);
+      if(exist == 0) {
+        char *archthinbin = malloc(strlen(ftmpdir)+strlen(arch)+11);
+        strcat(archthinbin, ftmpdir);
+        strcat(archthinbin, "/biberthin_");
+        strcat(archthinbin, arch);
+        char *lipocmd = malloc(strlen(arch)+strlen(archthinbin)+strlen(my_prog)+40);
```

Comment:
Please use fork + execve to run lipo, no need to cobble a command line together, just do
```
char* lipo_argv[] = { "/usr/bin/lipo", "-extract_family", arch, "-output", path, my_prog, NULL }:
```

Reply:
Done.